File: algorithm/baekjoon/codeplus/first-solve/src/main/java/com/keunwon/algorithm/graph/Problem2667.java
package com.keunwon.algorithm.graph;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Problem2667 {
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static final int[][] moves = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    private static final List<Integer> result = new ArrayList<>();

    private static int N;
    private static int[][] map;
    private static boolean[][] visited;

    private static int count = 0;

    public static void main(String[] args) throws IOException {
        input();

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (isGo(i, j)) {
                    count = 1;
                    dfs(i, j);
                    result.add(count);
                }
            }
        }

        Collections.sort(result);

        System.out.println(result.size());
        result.forEach(System.out::println);
    }

    private static void dfs(int x, int y) {
        visited[x][y] = true;

        for (int[] move : moves) {
            final int nx = x + move[0];
            final int ny = y + move[1];

            if (validMap(nx, ny) && isGo(nx, ny)) {
                dfs(nx, ny);
                count++;
            }
        }
    }

    private static boolean validMap(int x, int y) {
        return 0 <= x && x < N && 0 <= y && y < N;
    }

    private static boolean isGo(int x, int y) {
        return map[x][y] == 1 && !visited[x][y];
    }

    private static void input() throws IOException {
        N = Integer.parseInt(br.readLine());
        map = new int[N][N];
        visited = new boolean[N][N];

        for (int i = 0; i < N; i++) {
            String s = br.readLine();

            for (int j = 0; j < N; j++) {
                map[i][j] = s.charAt(j) - '0';
            }
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem2667.kt
package com.keunwon.algorithm.baekjoon

import java.util.*

/**
 * Title: 단지번호붙이기
 * Level: 실버-1
 **/
class Problem2667 {
    private val moves = arrayOf(-1 to 0, 0 to 1, 1 to 0, 0 to -1)

    fun solution(map: Array<IntArray>): List<Int> {
        val visited = Array(map.size) { BooleanArray(map[0].size) }
        val answer = mutableListOf<Int>()

        for (i in map.indices) {
            for (j in map[i].indices) {
                if (map[i][j] == 1 && !visited[i][j]) {
                    val count = bfs(Node(i, j), map, visited)
                    answer.add(count)
                }
            }
        }
        return answer.sorted()
    }

    private fun bfs(start: Node, map: Array<IntArray>, visited: Array<BooleanArray>): Int {
        val queue = LinkedList<Node>()
        var count = 0

        queue.offer(start)
        visited[start.r][start.c] = true

        while (queue.isNotEmpty()) {
            val cur = queue.poll()
            count++

            for ((mr, mc) in moves) {
                val nr = cur.r + mr
                val nc = cur.c + mc

                if (nr !in map.indices || nc !in map[0].indices) continue
                if (map[nr][nc] == 0 || visited[nr][nc]) continue

                visited[nr][nc] = true
                queue.offer(Node(nr, nc))
            }
        }
        return count
    }

    private data class Node(val r: Int, val c: Int)
}

fun main() = System.`in`.bufferedReader().use { br ->
    System.out.bufferedWriter().use { bw ->
        val n = br.readLine().toInt()
        val map = Array(n) {
            br.readLine().toCharArray().map { it.digitToInt() }.toIntArray()
        }

        val list = Problem2667().solution(map)
        bw.write("${list.size}")
        bw.newLine()

        list.forEach {
            bw.write("$it")
            bw.newLine()
        }

        bw.flush()
    }
}
